autoload/ale/handlers/rust.vim

```diff
@@ -50,7 +50,7 @@ function! ale#handlers#rust#HandleRustErrorsForFile(buffer, full_filename, lines
 
             if (
             \   l:span.is_primary
-            \   && (l:span_filename ==# l:filename || l:span_filename ==# '<anon>')
+            \   && (a:full_filename =~ (l:span.file_name . '$') || l:span_filename ==# '<anon>')
```

Comment:
You may have to add `silent file ...` to the tests after changing this, so the tests run with the buffer set to a filename we should match against.